src/frontend/src/binder/mod.rs

```diff
@@ -82,12 +92,28 @@ impl Binder {
 
     fn push_context(&mut self) {
         let new_context = std::mem::take(&mut self.context);
-        self.upper_contexts.push(new_context);
+        let new_table_contexts = std::mem::take(&mut self.upper_table_contexts);
+        self.upper_subquery_contexts
+            .push((new_context, new_table_contexts));
     }
 
     fn pop_context(&mut self) {
-        let old_context = self.upper_contexts.pop();
-        self.context = old_context.unwrap();
+        let (old_context, old_table_contexts) = self.upper_subquery_contexts.pop().unwrap();
+        self.context = old_context;
+        self.upper_table_contexts = old_table_contexts;
+    }
+
+    fn push_table_context(&mut self) {
+        let new_context = std::mem::take(&mut self.context);
+        self.upper_table_contexts.push(new_context);
+    }
+
+    fn pop_and_merge_table_context(&mut self) {
+        let mut old_context = self.upper_table_contexts.pop().unwrap();
+        old_context
+            .merge_context(self.context.clone())
+            .expect("could not merge contexts");
```

Comment:
should not panic here?

Reply:
Maybe all of the functions here need to return a `Result<()>`